### PR TITLE
gh-126639: Add ResourceWarning to NamedTemporaryFile

### PR DESCRIPTION
**Pull request by graingert**
* Issue: gh-126639

### CODE REVIEW

**Comment by graingert** [comment]
Ok this doesn't seem to break anything! I'll need to add a test though

File: Misc/NEWS.d/next/Library/2024-11-11-07-56-03.gh-issue-126639.AmVSt-.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Add :exc:`ResourceWarning` to :class:`tempfile.NamedTemporaryFile`
```
Comment (graingert) [inline Misc/NEWS.d/next/Library/2024-11-11-07-56-03.gh-issue-126639.AmVSt-.rst]:
```suggestion
:class:`tempfile.NamedTemporaryFile` will now issue a :exc:`ResourceWarning` when it is finalized by the garbage collector before without being explicitly closed.
```

File: Lib/test/test_tempfile.py [hunk 2]
```diff
@@ -1116,11 +1116,14 @@ def my_func(dir):
...
             # f.close()
             return tmp_name
-        # Make sure that the garbage collector has finalized the file object.
-        gc.collect()
         dir = tempfile.mkdtemp()
         try:
-            tmp_name = my_func(dir)
+            with self.assertWarnsRegex(
+                ResourceWarning,
+                msg=r"Implicitly cleaning up <_TemporaryFileWrapper file=.*>",
```
Comment (graingert) [inline Lib/test/test_tempfile.py]:
```suggestion
                expected_warning=ResourceWarning,
                expected_regex=r"Implicitly cleaning up <_TemporaryFileWrapper file=.*>",
```

File: Lib/tempfile.py [hunk 3]
```diff
@@ -483,8 +494,17 @@ class _TemporaryFileWrapper:
...
         self.file = file
         self.name = name
-        self._closer = _TemporaryFileCloser(file, name, delete,
-                                            delete_on_close)
+        self._closer = _TemporaryFileCloser(
+            file,
+            name,
+            delete,
+            delete_on_close,
+            warn_message="Implicitly cleaning up {self!r}",
```
Comment (graingert) [inline Lib/tempfile.py]:
```suggestion
            warn_message=f"Implicitly cleaning up {self!r}",
```

File: Misc/NEWS.d/next/Library/2024-11-11-07-56-03.gh-issue-126639.AmVSt-.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+:class:`tempfile.NamedTemporaryFile` will now issue a :exc:`ResourceWarning` when it is finalized by the garbage collector before without being explicitly closed.
```
Comment (graingert) [inline Misc/NEWS.d/next/Library/2024-11-11-07-56-03.gh-issue-126639.AmVSt-.rst]:
```suggestion
:class:`tempfile.NamedTemporaryFile` will now issue a :exc:`ResourceWarning` when it is finalized by the garbage collector without being explicitly closed.
```

File: Lib/test/test_tempfile.py [hunk 5]
```diff
@@ -1116,11 +1116,14 @@ def my_func(dir):
             # Testing extreme case, where the file is not explicitly closed
             # f.close()
             return tmp_name
-        # Make sure that the garbage collector has finalized the file object.
-        gc.collect()
```
Comment (graingert) [inline Lib/test/test_tempfile.py:1116]: this collection didn't do anything because it ran before creating the object it was supposed to collect

**Comment by nineteendo** [comment]
You can add suggestions to batch from the files tab to apply multiple suggestions in a single commit.

**Comment by graingert** [comment]
> TIP: You can add suggestions to batch from the files tab to apply multiple suggestions in a single commit.

Oh yeah I know, but I found the issues with some time apart so they didn't exist together to be batched

**Review by asvetlov — APPROVED** [review]

**Review by kumaraditya303 — APPROVED** [review]

**Comment by graingert** [comment]
Now the question remains, is this a bug or a feature?

**Comment by kumaraditya303** [comment]
I would treat it as feature, we don't add any new warnings in stable releases.

**Comment by graingert** [comment]
Fair

**Pull request merged**